src/docs/rfcs/018-declarative-vdu.md

```diff
@@ -0,0 +1,2249 @@
+  the root of the input document, such that a client could locate the
+  non-matching value by evaluating `path.reduce((val, key) => val[key], doc)`
+- `type`: the name of the matching operator that failed, e.g. `"eq"`, `"in"`,
+  `"type"`, etc.
+- `params`: an array of any other values that the operator used to determine the
+  match result. For most operators this would just be the expected value.
+
+An example of a failure object:
+
+    {
+      "path": ["$secObj", "members", "names", 1],
+      "type": "eq",
+      "params": ["Alice"]
+    }
+
+A client should be able to construct useful user-facing error messages from the
+information in these failure objects such that the user could correct any
+mistakes in their input.
+
+An Extended Mango expression is considered to match a given input if its
+evaluation returns an empty list.
+
+To produce the `path` field, the `match` function will need to track the path it
+used to reach the current value. This can be done by adding a "match context" to
+its list of parameters that tracks this information along with other things.
+This will also be needed to handle negation and relative `$data` references.
+
+
+### General evaluation
+
+Since `match` currently just returns `true` or `false`, the `mango_selector`
+module can implement certain operations using "short-circuit" semantics. For
+example, `$and` can be implemented by checking each of its sub-selectors, and
+returning `false` as soon as a single one of them returns `false`. Likewise
+`$or` can return `true` as soon as single sub-selector returns `true`.
+
+For VDUs, we want to return a complete list of match failures to the client, so
+some compound operators must evaluate all their inputs completely, without
+short-circuiting. Specifically:
+
+- `$and` should evaluate all its sub-selectors and return the combined list of
+  failures produced by any of them.
+- `$or` may return an empty list as soon as any of its sub-selectors returns an
+  empty list. If all its sub-selectors return non-empty failure lists, it should
+  return a combined list of all failures.
+- `$nor` should be translated as described under "Negation" below, before the
+  expression is evaluated.
+- `$allMatch` should evaluate all list items against the sub-selector and return
+  a list of all failures produced by any of the items. It must only return an
+  empty list of all items produce an empty list.
+- `$elemMatch` may return an empty list as soon as an item is found that
+  produces an empty list for the sub-selector. If no item does so, the combined
+  list of failures from all items should be returned.
+
+For normal object matchers, `{ "a": X, "b": Y }` should have the same behaviour
+as `{ "$and": [{ "a": X }, { "b": Y }] }`. That is, all the fields in the object
+should be checked, rather than returning `false` as soon as a single field does
+not match, and all failures from all fields should be returned to the caller.
+
+
+### `$if`/`$then`/`$else`
```

Comment:
To me, aesthetically if/then/else seem a bit out of place in what looks like a declarative query language. It makes me think of an imperative language like Javascript or Python. Especially as the ifs/thens can be translated to a bunch of ANDs and ORs user could still build the same logic with those, and it be nice to minimize change to our custom selector syntax and grammar.